src/Generator/Generators/CSharp/CSharpMarshal.cs

```diff
@@ -382,16 +385,26 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
             if (!VisitType(array, quals))
                 return false;
 
+            Class @class;
             switch (array.SizeType)
             {
                 case ArrayType.ArraySize.Constant:
                     var supportBefore = Context.SupportBefore;
                     supportBefore.WriteLine("if ({0} != null)", Context.ArgName);
                     supportBefore.WriteStartBraceIndent();
+                    if (array.Type.Desugar().TryGetClass(out @class) && @class.IsRefType)
+                    {
+                        supportBefore.WriteLine("if (value.Length != {0})", array.Size);
+                        supportBefore.WriteLineIndent("throw new ArrayTypeMismatchException(\"The provided array's dimentions don't match the requred size.\");");
```

Comment:
The message is OK, I think, however we should choose a better exception type. Let me check...